schematic/schemas/json_schema_generator.py

```diff
@@ -0,0 +1,747 @@
+
+    def update_nodes_to_process(self, nodes: list[str]) -> None:
+        """Updates the nodes to process with the input nodes
+
+        Arguments:
+            nodes: Nodes to add
+        """
+        self.nodes_to_process += nodes
+
+    def update_processed_nodes_with_current_node(self) -> None:
+        """Adds the current node to the list of processed nodes"""
+        if self.current_node is None:
+            raise ValueError("Current node is None")
+        self.processed_nodes.append(self.current_node)
+
+
+class JSONSchemaGenerator:  # pylint: disable=too-few-public-methods
+    "Data Model Json Schema"
+
+    def __init__(
+        self,
+        jsonld_path: str,
+        graph: nx.MultiDiGraph,
+    ) -> None:
+        self.jsonld_path = jsonld_path
+        self.graph = graph
+        self.dmge = DataModelGraphExplorer(self.graph)
+        self.dmr = DataModelRelationships()
+        self.rel_dict = self.dmr.relationships_dictionary
+
+    def create_json_schema(
+        self,
+        datatype: str,
+        schema_name: str,
+        schema_path: Union[str, None] = None,
+        write_schema: bool = True,
+    ) -> dict[str, Any]:
+        """
+        Consolidated method that aims to gather dependencies and value constraints across terms
+        / nodes in a schema.org schema and store them in a jsonschema /JSON Schema schema.
+
+        It does so for any given node in the schema.org schema (recursively) using the given
+          node as starting point in the following manner:
+        1) Find all the nodes / terms this node depends on (which are required as
+          "additional metadata" given this node is "required").
+        2) Find all the allowable metadata values / nodes that can be assigned to a particular
+          node (if such a constraint is specified on the schema).
+
+        Arguments:
+            datatype: the datatype to create the schema for.
+              Its node is where we can start recursive dependency traversal
+              (as mentioned above).
+            schema_name: Name assigned to JSON-LD schema (to uniquely identify it via URI
+              when it is hosted on the Internet).
+            schema_path: Where to save the JSON Schema file
+            write_schema: whether or not to write the schema as a json file
+
+        Returns:
+            JSON Schema as a dictionary.
+        """
+
+        # Gets the dependency nodes of the source node. These will be first nodes processed.
+        root_dependencies = self.dmge.get_adjacent_nodes_by_relationship(
+            node_label=datatype,
+            relationship=self.rel_dict["requiresDependency"]["edge_key"],
+        )
+        # if root_dependencies is empty it means that a class with name 'source_node' exists
+        # in the schema, but it is not a valid component
+        if not root_dependencies:
+            raise ValueError(f"'{datatype}' is not a valid component in the schema.")
```

Comment:
Should this exception be expanded to include the notes from your comment above?

Reply:
@BryanFauble TBH this comment is leftover from the previous class. I had trouble understanding what it means. @linglp @SageGJ Do you know why a class would have no root dependencies? Is that a malformed data-model?